File: server/cmd/api/api/webmcp.go
package api

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/kernel/kernel-images/server/lib/logger"
	"github.com/kernel/kernel-images/server/lib/oapi"
	"github.com/kernel/kernel-images/server/lib/webmcpclient"
)

const (
	defaultWebMCPInvocationTimeout = 60 * time.Second
	maxWebMCPInvocationTimeoutSec  = 120
	maxWebMCPInputBytes            = 1 << 20
	maxWebMCPRequestBytes          = maxWebMCPInputBytes + (4 << 10)
)

func (s *ApiService) GetWebMCPTools(ctx context.Context, _ oapi.GetWebMCPToolsRequestObject) (oapi.GetWebMCPToolsResponseObject, error) {
	tools, err := s.webmcp.Tools(ctx)
	if err != nil {
		if errors.Is(err, webmcpclient.ErrNoPageTarget) {
			return oapi.GetWebMCPTools404JSONResponse{NotFoundErrorJSONResponse: oapi.NotFoundErrorJSONResponse{Message: err.Error()}}, nil
		}
		logger.FromContext(ctx).Error("failed to discover WebMCP tools", "err", err)
		return oapi.GetWebMCPTools500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "failed to discover WebMCP tools"}}, nil
	}

	responseTools := make([]oapi.WebMCPTool, 0, len(tools))
	for _, tool := range tools {
		inputSchema := tool.InputSchema
		if inputSchema == nil {
			inputSchema = make(map[string]any)
		}
		responseTool := oapi.WebMCPTool{
			ToolRef:     tool.Ref,
			Name:        tool.Name,
			Description: tool.Description,
			InputSchema: inputSchema,
			Source: oapi.WebMCPToolSource{
				WindowId:  tool.Source.WindowID,
				TabId:     tool.Source.TabID,
				PageTitle: tool.Source.PageTitle,
				PageUrl:   tool.Source.PageURL,
			},
		}
		if tool.Source.Frame != nil {
			responseTool.Source.Frame = &oapi.WebMCPToolFrame{
				FrameId: tool.Source.Frame.FrameID,
				Url:     tool.Source.Frame.URL,
			}
		}
		if tool.Annotations != nil {
			responseTool.Annotations = &oapi.WebMCPToolAnnotations{
				ReadOnly:         tool.Annotations.ReadOnly,
				UntrustedContent: tool.Annotations.UntrustedContent,
				Consequential:    tool.Annotations.Consequential,
				Autosubmit:       tool.Annotations.Autosubmit,
			}
		}
		responseTools = append(responseTools, responseTool)
	}
	return oapi.GetWebMCPTools200JSONResponse{Tools: responseTools}, nil
}

func (s *ApiService) InvokeWebMCPTool(ctx context.Context, request oapi.InvokeWebMCPToolRequestObject) (oapi.InvokeWebMCPToolResponseObject, error) {
	if request.Body == nil {
		return oapi.InvokeWebMCPTool400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "request body is required"}}, nil
	}
	toolRefLength := utf8.RuneCountInString(request.Body.ToolRef)
	if toolRefLength < 1 || toolRefLength > 128 {
		return oapi.InvokeWebMCPTool400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "tool_ref must be between 1 and 128 characters"}}, nil
	}
	if request.Body.Input == nil {
		return oapi.InvokeWebMCPTool400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "input is required and must be an object"}}, nil
	}
	inputJSON, err := json.Marshal(request.Body.Input)
	if err != nil || len(inputJSON) > maxWebMCPInputBytes {
		return oapi.InvokeWebMCPTool400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "input must be valid JSON no larger than 1 MiB"}}, nil
	}
	timeout := defaultWebMCPInvocationTimeout
	if request.Body.TimeoutSec != nil {
		if *request.Body.TimeoutSec < 1 || *request.Body.TimeoutSec > maxWebMCPInvocationTimeoutSec {
			return oapi.InvokeWebMCPTool400JSONResponse{BadRequestErrorJSONResponse: oapi.BadRequestErrorJSONResponse{Message: "timeout_sec must be between 1 and 120"}}, nil
		}
		timeout = time.Duration(*request.Body.TimeoutSec) * time.Second
	}
	invokeCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	result, err := s.webmcp.Invoke(invokeCtx, request.Body.ToolRef, request.Body.Input)
	if err != nil {
		switch {
		case errors.Is(err, webmcpclient.ErrToolNotFound):
			return oapi.InvokeWebMCPTool404JSONResponse{NotFoundErrorJSONResponse: oapi.NotFoundErrorJSONResponse{Message: "WebMCP tool is no longer available; discover tools again"}}, nil
		case errors.Is(err, webmcpclient.ErrOutcomeUnknown):
			failure := oapi.WebMCPInvocationFailure{
				Code:    oapi.OutcomeUnknown,
				Message: "the invocation started, but its final outcome could not be observed; do not retry automatically",
			}
			failure.InvocationId = nonEmptyString(result.InvocationID)
			return oapi.InvokeWebMCPTool504JSONResponse(failure), nil
		default:
			logger.FromContext(ctx).Error("failed to invoke WebMCP tool", "err", err)
			return oapi.InvokeWebMCPTool500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "failed to invoke WebMCP tool"}}, nil
		}
	}

	status := oapi.WebMCPInvocationResultStatus(strings.ToLower(result.Status))
	if !status.Valid() {
		logger.FromContext(ctx).Error("WebMCP tool returned unknown status", "status", result.Status)
		return oapi.InvokeWebMCPTool500JSONResponse{InternalErrorJSONResponse: oapi.InternalErrorJSONResponse{Message: "WebMCP tool returned an unknown status"}}, nil
	}
	response := oapi.InvokeWebMCPTool200JSONResponse{
		InvocationId: result.InvocationID,
		Status:       status,
		Output:       result.Output,
	}
	response.ErrorText = nonEmptyString(result.ErrorText)
	return response, nil
}

func nonEmptyString(value string) *string {
	if value == "" {
		return nil
	}
	return &value
}

File: server/runtime/webmcp.ts
export interface WebMCPToolFrame {
  frame_id: number;
  url: string;
}

export interface WebMCPToolSource {
  window_id: number;
  tab_id: number;
  page_title: string;
  page_url: string;
  frame: WebMCPToolFrame | null;
}

export interface WebMCPTool {
  tool_ref: string;
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
  annotations?: Record<string, boolean>;
  source: WebMCPToolSource;
}

export interface WebMCPInvocationResult {
  invocation_id: string;
  status: 'completed' | 'canceled' | 'error' | 'awaiting_submission';
  output?: unknown;
  error_text?: string;
}

export interface WebMCPInvokeOptions {
  timeoutSec?: number;
}

export interface WebMCPClient {
  listTools(): Promise<WebMCPTool[]>;
  invokeTool(
    toolRef: string,
    input?: Record<string, unknown>,
    options?: WebMCPInvokeOptions,
  ): Promise<WebMCPInvocationResult>;
}

export class WebMCPRequestError extends Error {
  readonly statusCode: number;
  readonly code?: string;
  readonly invocationId?: string;
  readonly body: unknown;

  constructor(statusCode: number, body: unknown) {
    const fields = isRecord(body) ? body : {};
    const code = typeof fields.code === 'string' ? fields.code : undefined;
    const invocationId =
      typeof fields.invocation_id === 'string' ? fields.invocation_id : undefined;
    const detail =
      typeof fields.message === 'string'
        ? fields.message
        : `request failed with status ${statusCode}`;
    // The Playwright daemon only forwards error.message, so the message itself
    // must identify this as a WebMCP failure and carry the code and invocation.
    const parts = [`WebMCP ${code ?? 'error'}`];
    if (invocationId) parts.push(`invocation ${invocationId}`);
    super(`${parts.join(', ')}: ${detail}`);
    this.name = 'WebMCPRequestError';
    this.statusCode = statusCode;
    this.code = code;
    this.invocationId = invocationId;
    this.body = body;
  }
}

interface WebMCPClientOptions {
  apiBaseUrl: string;
  signal?: AbortSignal;
  fetchImpl?: typeof fetch;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

async function responseBody(response: Response): Promise<unknown> {
  const text = await response.text();
  if (text === '') return null;
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

export function createWebMCPClient({
  apiBaseUrl,
  signal,
  fetchImpl = fetch,
}: WebMCPClientOptions): WebMCPClient {
  const baseUrl = apiBaseUrl.replace(/\/+$/, '');

  const request = async (path: string, init?: RequestInit): Promise<unknown> => {
    const response = await fetchImpl(`${baseUrl}${path}`, {...init, signal});
    const body = await responseBody(response);
    if (!response.ok) throw new WebMCPRequestError(response.status, body);
    return body;
  };

  const client: WebMCPClient = {
    async listTools() {
      const body = await request('/webmcp/tools');
      if (!isRecord(body) || !Array.isArray(body.tools)) {
        throw new Error('WebMCP tools response is invalid');
      }
      return body.tools as WebMCPTool[];
    },

    async invokeTool(toolRef, input = {}, options = {}) {
      const payload: Record<string, unknown> = {tool_ref: toolRef, input};
      if (options.timeoutSec !== undefined) payload.timeout_sec = options.timeoutSec;
      const body = await request('/webmcp/invoke', {
        method: 'POST',
        headers: {'content-type': 'application/json'},
        body: JSON.stringify(payload),
      });
      if (
        !isRecord(body) ||
        typeof body.invocation_id !== 'string' ||
        typeof body.status !== 'string'
      ) {
        throw new Error('WebMCP invocation response is invalid');
      }
      return body as WebMCPInvocationResult;
    },
  };

  return Object.freeze(client);
}
